translib/translib.go

```diff
@@ -478,8 +495,293 @@ func Get(req GetRequest) (GetResponse, error) {
 }
 
 func Action(req ActionRequest) (ActionResponse, error) {
+	var payload []byte
 	var resp ActionResponse
+	path := req.Path
+
+
+	log.Info("Received Action request for path = ", path)
+
+	app, appInfo, err := getAppModule(path, req.ClientVersion)
+
+	if err != nil {
+		resp = ActionResponse{Payload: payload, ErrSrc: ProtoErr}
+		return resp, err
+	}
+
+	aInfo := *appInfo
+
+	aInfo.isNative = true
+
+	err = appInitialize(app, &aInfo, path, &req.Payload, nil, GET)
+
+	if err != nil {
+		resp = ActionResponse{Payload: payload, ErrSrc: AppErr}
+		return resp, err
+	}
+
+    writeMutex.Lock()
+    defer writeMutex.Unlock()
+
+	isGetCase := false
+	dbs, err := getAllDbs(isGetCase)
+
+	if err != nil {
+		resp = ActionResponse{Payload: payload, ErrSrc: ProtoErr}
+		return resp, err
+	}
+
+	defer closeAllDbs(dbs[:])
+
+	err = (*app).translateAction(dbs)
+
+	if err != nil {
+		resp = ActionResponse{Payload: payload, ErrSrc: AppErr}
+		return resp, err
+	}
+
+	resp, err = (*app).processAction(dbs)
+
+	return resp, err
+}
+
+func Bulk(req BulkRequest) (BulkResponse, error) {
 	var err error
+	var keys []db.WatchKeys
+	var errSrc ErrSource
+
+	delResp := make([]SetResponse, len(req.DeleteRequest))
+	replaceResp := make([]SetResponse, len(req.ReplaceRequest))
+	updateResp := make([]SetResponse, len(req.UpdateRequest))
+	createResp := make([]SetResponse, len(req.CreateRequest))
+
+	resp := BulkResponse{DeleteResponse: delResp,
+		ReplaceResponse: replaceResp,
+		UpdateResponse: updateResp,
+		CreateResponse: createResp}
+
+	writeMutex.Lock()
+	defer writeMutex.Unlock()
+
+	isWriteDisabled := false
+	d, err := db.NewDB(getDBOptions(db.ConfigDB, isWriteDisabled))
+
+	if err != nil {
+		return resp, err
+	}
+
+	defer d.DeleteDB()
+
+	//Start the transaction without any keys or tables to watch will be added later using AppendWatchTx
+	err = d.StartTx(nil, nil)
+
+	if err != nil {
+        return resp, err
+    }
+
+	for i := range req.DeleteRequest {
+		path := req.DeleteRequest[i].Path
+
+		log.Info("Delete request received with path =", path)
+
+		app, appInfo, err := getAppModule(path, req.DeleteRequest[i].ClientVersion)
+
+		if err != nil {
+			errSrc = ProtoErr
```

Comment:
Please avoid goto.
You may set errSrc at the start, as it is of significance only when err != nil.
Instead of checking err != nill, check err = nil, for next command execution.

Reply:
Renuka, We have used goto throughout the code. Even in the checked in code we have goto. May I know the reason to avoid goto. I think we need to leave this to programmers preference.

Reply:
"goto" is generally not warranted in a well structured code. So in some places, they strictly block "goto", so you write better code :-). 

Even here, instead of checking 'if err !=nil',  if you switch to 'err == nil', you avoid goto and the code is better readable. Plus, if I am debugging a code, with goto statements, it is pretty tough, where as if not, I am certain to hit my break point and by looking at local variables, I can gauge the point of failure.

It is not a simple task, to go back and change all over. At the least, let us avoid, from here on.

Reply:
Let us  keep it as a requirement, from here on to avoid "goto" for new code.